src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs

```diff
@@ -63,12 +63,24 @@ public virtual void Load()
             string? parentPath)
         {
             var results = new List<string>();
+            results.AddRange(GetChildKeysInternal(parentPath));
+            results.AddRange(earlierKeys);
+            results.Sort(ConfigurationKeyComparer.Comparison);
+            return results;
+        }
 
+        /// <summary>
+        /// Returns the list of keys that this provider has.
+        /// </summary>
+        /// <param name="parentPath">The path for the parent IConfiguration.</param>
+        /// <returns>The list of keys for this provider.</returns>
+        internal IEnumerable<string> GetChildKeysInternal(string? parentPath)
+        {
             if (parentPath is null)
             {
                 foreach (KeyValuePair<string, string?> kv in Data)
                 {
-                    results.Add(Segment(kv.Key, 0));
+                    yield return Segment(kv.Key, 0);
```

Comment:
How do the benchmarks compare if we continue adding to a list instead of switching to using iterators? I can see iterators being better particularly for large numbers of keys, but I'm curious.

It might also be interesting to have a scneario with 8 providers and closer to 100 keys per provider.

Reply:
Sure. I can benchmark it tonight. Changing from Aggregate to SelectAll is essential for this optimization. Using list or iterator should have similar result.

Reply:
Using iterator (this PR):
| Method | ConfigurationProvidersCount | KeysCountPerProvider |     Mean |     Error |    StdDev |   Median |      Min |      Max |   Gen 0 |   Gen 1 | Allocated |
|------- |---------------------------- |--------------------- |---------:|----------:|----------:|---------:|---------:|---------:|--------:|--------:|----------:|
|    Get |                           8 |                  100 | 1.651 ms | 0.0320 ms | 0.0355 ms | 1.641 ms | 1.601 ms | 1.722 ms | 97.2222 | 13.8889 |  420.2 KB |

Using list (change to the code below):

| Method | ConfigurationProvidersCount | KeysCountPerProvider |     Mean |     Error |    StdDev |   Median |      Min |      Max |    Gen 0 |   Gen 1 | Allocated |
|------- |---------------------------- |--------------------- |---------:|----------:|----------:|---------:|---------:|---------:|---------:|--------:|----------:|
|    Get |                           8 |                  100 | 1.704 ms | 0.0223 ms | 0.0198 ms | 1.702 ms | 1.682 ms | 1.743 ms | 104.1667 | 10.4167 | 471.01 KB |


```
        internal IEnumerable<string> GetChildKeysInternal(string? parentPath)
        {
            var res = new List<string>();
            if (parentPath is null)
            {
                foreach (KeyValuePair<string, string?> kv in Data)
                {
                    res.Add(Segment(kv.Key, 0));
                }
            }
            else
            {
                Debug.Assert(ConfigurationPath.KeyDelimiter == ":");

                foreach (KeyValuePair<string, string?> kv in Data)
                {
                    if (kv.Key.Length > parentPath.Length &&
                        kv.Key.StartsWith(parentPath, StringComparison.OrdinalIgnoreCase) &&
                        kv.Key[parentPath.Length] == ':')
                    {
                        res.Add(Segment(kv.Key, parentPath.Length + 1));
                    }
                }
            }

            return res;
        }
```

Reply:
Changed back to list